tests/late_joiners.py

```diff
@@ -0,0 +1,216 @@
+    if suspend:
+        # We want to suspend the nodes' process so we need to initiate a new timer to wake it up eventually
+        node.suspend()
+        next_timeout = random.uniform(2 * election_timeout, 3 * election_timeout)
+        LOG.info(f"New timer set for node {node.node_id} is {next_timeout} seconds")
+        t = Timer(next_timeout, timeout, args=[node, False, 0])
+        t.start()
+    else:
+        node.resume()
+
+
+def run(args):
+    hosts = ["localhost", "localhost", "localhost"]
+
+    with infra.ccf.network(
+        hosts, args.build_dir, args.debug_nodes, args.perf_nodes, pdb=args.pdb
+    ) as network:
+        network.start_and_join(args)
+        first_node, backups = network.find_nodes()
+
+        term_info = {}
+        long_msg = "X" * (2 ** 14)
+
+        nodes_to_kill = []
+        for b in backups:
+            nodes_to_kill.append(b)
+        nodes_to_keep = [first_node]
+
+        # first timer determines after how many seconds each node will be suspended
+        if not args.skip_suspension:
+            timeouts = []
+            t = random.uniform(1, 10)
+            LOG.info(f"Initial timer for node {first_node.node_id} is {t} seconds...")
+            timeouts.append((t, first_node))
+            for backup in backups:
+                t = random.uniform(1, 10)
+                LOG.info(f"Initial timer for node {backup.node_id} is {t} seconds...")
+                timeouts.append((t, backup))
+
+            for t, node in timeouts:
+                tm = Timer(t, timeout, args=[node, True, args.election_timeout / 1000],)
+                tm.start()
+
+        with first_node.node_client() as mc:
+            check_commit = infra.checker.Checker(mc)
+            check = infra.checker.Checker()
+
+            clients = []
+            with contextlib.ExitStack() as es:
+                LOG.info("Write messages to nodes using round robin")
+                clients.append(es.enter_context(first_node.user_client(format="json")))
+                for backup in backups:
+                    clients.append(es.enter_context(backup.user_client(format="json")))
+                node_id = 0
+                for id in range(1, TOTAL_REQUESTS):
+                    if id == 1:
+                        LOG.info(
+                            "Adding another node after f = 0 but before we need to send append entries"
+                        )
+                        # check that a new node can catch up naturally
+                        new_node = network.create_and_trust_node(
+                            lib_name=args.package, host="localhost", args=args,
+                        )
+                        assert new_node
+                        nodes_to_keep.append(new_node)
+                    node_id += 1
+                    c = clients[node_id % len(clients)]
+                    try:
+                        resp = c.rpc("LOG_record", {"id": id, "msg": long_msg})
+                    except Exception:
```

Comment:
The exception should be more specific here. I guess it will be some kind of connection error? The exception should also be the same for every client type use (FTCP, Requests or CURL).